addons/hr_holidays/models/resource.py

```diff
@@ -1,10 +1,98 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import fields, models
+from odoo import fields, models, api
+from datetime import datetime
 
 
 class CalendarLeaves(models.Model):
     _inherit = "resource.calendar.leaves"
 
     holiday_id = fields.Many2one("hr.leave", string='Leave Request')
+
+    def _reevaluate_leaves(self, leaves):
+        for leave in leaves:
+            if leave.state == 'validate':
+                leave.action_refuse()
+                leave.action_draft()
+                leave._compute_number_of_days()
+                if leave.number_of_days > 0.0:
+                    leave.action_confirm()
+                    leave.action_validate()
+            else:
+                leave._compute_number_of_days()
+
+            if leave.number_of_days == 0.0:
+                leave.force_cancel("A new public holiday completely overrides this leave")
+        return True
```

Comment:
It would be nice if we batched calls to `action_refuse`, `action_draft`, `action_confirm` and `action_validate` since they can be quite expensive functions (see `activity_update`) (unfortunately `_compute_number_of_days` is not _really_ batched)

Speaking of `activity_update` I also wonder if it wouldn't be better to add a context key to avoid it, since it is meant to send a notification to the time off manager to validate the leave (which had already been validated)
```suggestion
    def _reevaluate_leaves(self, leaves):
        if not leaves:
            return
        validated_leaves = leaves.filtered(lambda l: l.state == 'validate')
        validated_leaves.action_refuse()
        validates_leaves.action_draft()
        leaves._compute_number_of_days()
        leaves_to_validate = validated_leaves.filtered(lambda l: l.number_of_days > 0)
        leaves_to_validate.action_confirm()
        leaves_to_validate.action_validate()
        for leave in leaves:
            if leave.number_of_days == 0.0:
                leave.force_cancel(_("A new public holiday completely overrides this leave"))
        return True
```